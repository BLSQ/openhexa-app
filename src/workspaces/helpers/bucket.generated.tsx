import * as Types from '../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type GetFileDownloadUrlMutationVariables = Types.Exact<{
  input: Types.PrepareObjectDownloadInput;
}>;


export type GetFileDownloadUrlMutation = { __typename?: 'Mutation', prepareObjectDownload: { __typename?: 'PrepareObjectDownloadResult', success: boolean, downloadUrl?: any | null } };

export type DeleteBucketObjectMutationVariables = Types.Exact<{
  input: Types.DeleteBucketObjectInput;
}>;


export type DeleteBucketObjectMutation = { __typename?: 'Mutation', deleteBucketObject: { __typename?: 'DeleteBucketObjectResult', success: boolean, errors: Array<Types.DeleteBucketObjectError> } };

export type GetBucketUploadUrlMutationVariables = Types.Exact<{
  input: Types.PrepareObjectUploadInput;
}>;


export type GetBucketUploadUrlMutation = { __typename?: 'Mutation', prepareObjectUpload: { __typename?: 'PrepareObjectUploadResult', success: boolean, uploadUrl?: any | null } };

export type CreateBucketFolderMutationVariables = Types.Exact<{
  input: Types.CreateBucketFolderInput;
}>;


export type CreateBucketFolderMutation = { __typename?: 'Mutation', createBucketFolder: { __typename?: 'CreateBucketFolderResult', success: boolean, errors: Array<Types.CreateBucketFolderError>, folder?: { __typename?: 'BucketObject', key: string, name: string, type: Types.BucketObjectType } | null } };


export const GetFileDownloadUrlDocument = gql`
    mutation GetFileDownloadUrl($input: PrepareObjectDownloadInput!) {
  prepareObjectDownload(input: $input) {
    success
    downloadUrl
  }
}
    `;
export type GetFileDownloadUrlMutationFn = Apollo.MutationFunction<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>;

/**
 * __useGetFileDownloadUrlMutation__
 *
 * To run a mutation, you first call `useGetFileDownloadUrlMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useGetFileDownloadUrlMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [getFileDownloadUrlMutation, { data, loading, error }] = useGetFileDownloadUrlMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useGetFileDownloadUrlMutation(baseOptions?: Apollo.MutationHookOptions<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>(GetFileDownloadUrlDocument, options);
      }
export type GetFileDownloadUrlMutationHookResult = ReturnType<typeof useGetFileDownloadUrlMutation>;
export type GetFileDownloadUrlMutationResult = Apollo.MutationResult<GetFileDownloadUrlMutation>;
export type GetFileDownloadUrlMutationOptions = Apollo.BaseMutationOptions<GetFileDownloadUrlMutation, GetFileDownloadUrlMutationVariables>;
export const DeleteBucketObjectDocument = gql`
    mutation deleteBucketObject($input: DeleteBucketObjectInput!) {
  deleteBucketObject(input: $input) {
    success
    errors
  }
}
    `;
export type DeleteBucketObjectMutationFn = Apollo.MutationFunction<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>;

/**
 * __useDeleteBucketObjectMutation__
 *
 * To run a mutation, you first call `useDeleteBucketObjectMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useDeleteBucketObjectMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [deleteBucketObjectMutation, { data, loading, error }] = useDeleteBucketObjectMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useDeleteBucketObjectMutation(baseOptions?: Apollo.MutationHookOptions<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>(DeleteBucketObjectDocument, options);
      }
export type DeleteBucketObjectMutationHookResult = ReturnType<typeof useDeleteBucketObjectMutation>;
export type DeleteBucketObjectMutationResult = Apollo.MutationResult<DeleteBucketObjectMutation>;
export type DeleteBucketObjectMutationOptions = Apollo.BaseMutationOptions<DeleteBucketObjectMutation, DeleteBucketObjectMutationVariables>;
export const GetBucketUploadUrlDocument = gql`
    mutation GetBucketUploadUrl($input: PrepareObjectUploadInput!) {
  prepareObjectUpload(input: $input) {
    success
    uploadUrl
  }
}
    `;
export type GetBucketUploadUrlMutationFn = Apollo.MutationFunction<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>;

/**
 * __useGetBucketUploadUrlMutation__
 *
 * To run a mutation, you first call `useGetBucketUploadUrlMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useGetBucketUploadUrlMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [getBucketUploadUrlMutation, { data, loading, error }] = useGetBucketUploadUrlMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useGetBucketUploadUrlMutation(baseOptions?: Apollo.MutationHookOptions<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>(GetBucketUploadUrlDocument, options);
      }
export type GetBucketUploadUrlMutationHookResult = ReturnType<typeof useGetBucketUploadUrlMutation>;
export type GetBucketUploadUrlMutationResult = Apollo.MutationResult<GetBucketUploadUrlMutation>;
export type GetBucketUploadUrlMutationOptions = Apollo.BaseMutationOptions<GetBucketUploadUrlMutation, GetBucketUploadUrlMutationVariables>;
export const CreateBucketFolderDocument = gql`
    mutation CreateBucketFolder($input: CreateBucketFolderInput!) {
  createBucketFolder(input: $input) {
    success
    errors
    folder {
      key
      name
      type
    }
  }
}
    `;
export type CreateBucketFolderMutationFn = Apollo.MutationFunction<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>;

/**
 * __useCreateBucketFolderMutation__
 *
 * To run a mutation, you first call `useCreateBucketFolderMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useCreateBucketFolderMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [createBucketFolderMutation, { data, loading, error }] = useCreateBucketFolderMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useCreateBucketFolderMutation(baseOptions?: Apollo.MutationHookOptions<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>(CreateBucketFolderDocument, options);
      }
export type CreateBucketFolderMutationHookResult = ReturnType<typeof useCreateBucketFolderMutation>;
export type CreateBucketFolderMutationResult = Apollo.MutationResult<CreateBucketFolderMutation>;
export type CreateBucketFolderMutationOptions = Apollo.BaseMutationOptions<CreateBucketFolderMutation, CreateBucketFolderMutationVariables>;