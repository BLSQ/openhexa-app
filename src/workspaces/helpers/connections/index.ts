import { ConnectionType } from "graphql-types";
export { convertFieldsToInput } from "./utils";
export type { ConnectionForm, FieldForm } from "./utils";

import postgresql from "./postgresql";
import { ConnectionDefinition } from "./utils";
import custom from "./custom";
import s3 from "./s3";
import gcs from "./gcs";
import iaso from "./iaso";
import dhis2 from "./dhis2";

const Connections: { [key in ConnectionType]: ConnectionDefinition } = {
  [ConnectionType.Postgresql]: postgresql,
  [ConnectionType.Custom]: custom,
  [ConnectionType.S3]: s3,
  [ConnectionType.Gcs]: gcs,
  [ConnectionType.Iaso]: iaso,
  [ConnectionType.Dhis2]: dhis2,
};

export default Connections;
