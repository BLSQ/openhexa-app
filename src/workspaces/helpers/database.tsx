type lang = "R" | "PYTHON";

export const getReadTableSnippet = (tableName: string) => {
  return `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

pd.read_sql("SELECT * FROM ${tableName}", con=engine)`;
};

export const getUsageSnippet = (tableName: string, lang: lang) => {
  let text = "";
  switch (lang) {
    case "PYTHON":
      text = `import os
import pandas as pd
from sqlalchemy import create_engine

engine = create_engine(os.environ["WORKSPACE_DATABASE_URL"])

# Create sample dataframe
df = pd.DataFrame({"name": ["Jane", "John", "Tyler"], "age": [19, 17, 22]})

# Write data
df.to_sql("${tableName}", con=engine, if_exists="replace")`;
      break;
    case "R":
      text = `library(DBI)
  
con <- dbConnect(
    RPostgres::Postgres(),
    dbname = Sys.getenv("WORKSPACE_DATABASE_DB_NAME"),
    host = Sys.getenv("WORKSPACE_DATABASE_HOST"),
    port = Sys.getenv("WORKSPACE_DATABASE_PORT"),
    user = Sys.getenv("WORKSPACE_DATABASE_USERNAME"),
    password = Sys.getenv("WORKSPACE_DATABASE_PASSWORD")
)

dbWriteTable(con, "${tableName}", Data_fin, overwrite=TRUE)`;
      break;
  }
  return text;
};
