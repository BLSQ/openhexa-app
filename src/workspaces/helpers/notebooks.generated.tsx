import * as Types from '../../graphql-types';

import { gql } from '@apollo/client';
import * as Apollo from '@apollo/client';
const defaultOptions = {} as const;
export type LaunchNotebookServerMutationVariables = Types.Exact<{
  input: Types.LaunchNotebookServerInput;
}>;


export type LaunchNotebookServerMutation = { __typename?: 'Mutation', launchNotebookServer: { __typename?: 'LaunchNotebookServerResult', success: boolean, server?: { __typename?: 'NotebookServer', name: string, ready: boolean, url: string } | null } };


export const LaunchNotebookServerDocument = gql`
    mutation launchNotebookServer($input: LaunchNotebookServerInput!) {
  launchNotebookServer(input: $input) {
    success
    server {
      name
      ready
      url
    }
  }
}
    `;
export type LaunchNotebookServerMutationFn = Apollo.MutationFunction<LaunchNotebookServerMutation, LaunchNotebookServerMutationVariables>;

/**
 * __useLaunchNotebookServerMutation__
 *
 * To run a mutation, you first call `useLaunchNotebookServerMutation` within a React component and pass it any options that fit your needs.
 * When your component renders, `useLaunchNotebookServerMutation` returns a tuple that includes:
 * - A mutate function that you can call at any time to execute the mutation
 * - An object with fields that represent the current status of the mutation's execution
 *
 * @param baseOptions options that will be passed into the mutation, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options-2;
 *
 * @example
 * const [launchNotebookServerMutation, { data, loading, error }] = useLaunchNotebookServerMutation({
 *   variables: {
 *      input: // value for 'input'
 *   },
 * });
 */
export function useLaunchNotebookServerMutation(baseOptions?: Apollo.MutationHookOptions<LaunchNotebookServerMutation, LaunchNotebookServerMutationVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useMutation<LaunchNotebookServerMutation, LaunchNotebookServerMutationVariables>(LaunchNotebookServerDocument, options);
      }
export type LaunchNotebookServerMutationHookResult = ReturnType<typeof useLaunchNotebookServerMutation>;
export type LaunchNotebookServerMutationResult = Apollo.MutationResult<LaunchNotebookServerMutation>;
export type LaunchNotebookServerMutationOptions = Apollo.BaseMutationOptions<LaunchNotebookServerMutation, LaunchNotebookServerMutationVariables>;