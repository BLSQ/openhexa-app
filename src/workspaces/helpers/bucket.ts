import { gql } from "@apollo/client";
import { getApolloClient } from "core/helpers/apollo";
import {
  GetFileDownloadUrlMutation,
  GetFileDownloadUrlMutationVariables,
  DeleteBucketObjectMutation,
  DeleteBucketObjectMutationVariables,
  GetBucketUploadUrlMutation,
  GetBucketUploadUrlMutationVariables,
} from "./bucket.generated";

export async function getBucketObjectDownloadUrl(
  workspaceSlug: string,
  objectKey: string,
): Promise<string> {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetFileDownloadUrlMutation,
    GetFileDownloadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetFileDownloadUrl($input: PrepareObjectDownloadInput!) {
        prepareObjectDownload(input: $input) {
          success
          downloadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey,
      },
    },
  });

  if (data?.prepareObjectDownload?.success) {
    return data.prepareObjectDownload.downloadUrl as string;
  } else {
    throw new Error("Object cannot be downloaded");
  }
}

export async function downloadURL(url: string, target: string = "") {
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.target = target;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
}

export async function deleteBucketObject(
  workspaceSlug: string,
  objectKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    DeleteBucketObjectMutation,
    DeleteBucketObjectMutationVariables
  >({
    mutation: gql`
      mutation deleteBucketObject($input: DeleteBucketObjectInput!) {
        deleteBucketObject(input: $input) {
          success
          errors
        }
      }
    `,
    variables: { input: { workspaceSlug, objectKey } },
    update(cache) {
      const normalizedId = cache.identify({
        key: objectKey,
        __typename: "BucketObject",
      });

      cache.evict({ id: normalizedId });
      cache.gc();
    },
  });

  if (!data?.deleteBucketObject.success) {
    throw new Error("Unexpected error.");
  }
}

export async function getBucketObjectUploadUrl(
  workspaceSlug: string,
  key: string,
  contentType: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    GetBucketUploadUrlMutation,
    GetBucketUploadUrlMutationVariables
  >({
    mutation: gql`
      mutation GetBucketUploadUrl($input: PrepareObjectUploadInput!) {
        prepareObjectUpload(input: $input) {
          success
          uploadUrl
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        objectKey: key,
        contentType,
      },
    },
  });

  if (data?.prepareObjectUpload?.success) {
    return data.prepareObjectUpload.uploadUrl as string;
  } else {
    throw new Error("Object cannot be uploaded");
  }
}

export async function createBucketFolder(
  workspaceSlug: string,
  folderKey: string,
) {
  const client = getApolloClient();
  const { data } = await client.mutate({
    mutation: gql`
      mutation CreateBucketFolder($input: CreateBucketFolderInput!) {
        createBucketFolder(input: $input) {
          success
          errors
          folder {
            key
            name
            type
          }
        }
      }
    `,
    variables: {
      input: {
        workspaceSlug,
        folderKey,
      },
    },
  });

  if (!data?.createBucketFolder.success) {
    throw new Error("Unexpected error.");
  }

  return data.createBucketFolder.folder;
}
