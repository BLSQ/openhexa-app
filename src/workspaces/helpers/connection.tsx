import { gql } from "@apollo/client";
import { PlusCircleIcon, XMarkIcon } from "@heroicons/react/24/outline";
import Button from "core/components/Button";
import Checkbox from "core/components/forms/Checkbox";
import Field from "core/components/forms/Field";
import Title from "core/components/Title";
import { getApolloClient } from "core/helpers/apollo";
import { FormInstance } from "core/hooks/useForm";
import { Connection, ConnectionType } from "graphql-types";
import { useCallback, useEffect, useMemo } from "react";
import { useTranslation } from "react-i18next";
import {
  DeleteConnectionMutation,
  DeleteConnectionMutationVariables,
} from "./connection.generated";

type FieldForm = { secret?: boolean; code: string; value?: string | null } & {
  [key: string]: any;
};

export type ConnectionForm = {
  name: string;
  description?: string;
  slug?: string;
  type: ConnectionType | null;
  fields: FieldForm[];
};

const updateFormField = (
  form: FormInstance<ConnectionForm>,
  index: number,
  values: any
) => {
  const newFields = [...(form.formData.fields ?? [])];
  newFields.splice(index, 1, { ...newFields[index], ...values });
  form.setFieldValue("fields", newFields);
};

export function getUsageSnippets(
  connection: Pick<Connection, "slug" | "type"> & {
    fields: { code: string }[];
  }
) {
  const slugify = (...keys: string[]) =>
    keys.join("_").replace("-", "_").toUpperCase();

  switch (connection.type) {
    case ConnectionType.Dhis2:
      return [
        {
          lang: "python",
          code: `# Importing os module 
import os

DHIS2_URL = os.getenv("${slugify(connection.slug, "api_url")}")
DHIS2_USERNAME = os.getenv("${slugify(connection.slug, "username")}")
DHIS2_PASSWORD = os.getenv("${slugify(connection.slug, "password")}")
`,
        },
      ];
    case ConnectionType.Postgresql:
      return [
        {
          lang: "python",
          code: `# Importing os module 
import os

# import postgresql library
import psycopg2

# Create the connection to the database
conn = psycopg2.connect(
  database=os.getenv("${slugify(connection.slug, "db_name")}"),
  host=os.getenv("${slugify(connection.slug, "host")}"),
  user=os.getenv("${slugify(connection.slug, "user")}"),
  password=os.getenv("${slugify(connection.slug, "password")}"),
  port=os.getenv("${slugify(connection.slug, "port")}")
)

# Create a DB session
cursor = conn.cursor()
`,
        },
      ];
    case ConnectionType.S3:
      return [
        {
          lang: "python",
          code: `# importing os module 
import os

# Reading and writing to S3 using Pandas
import pandas as pd
import boto3

AWS_S3_BUCKET = os.getenv("${slugify(connection.slug, "bucket_name")}")
AWS_ACCESS_KEY_ID = os.getenv("${slugify(connection.slug, "access_key_id")}")
AWS_SECRET_ACCESS_KEY = os.getenv("${slugify(
            connection.slug,
            "access_key_secret"
          )}")
AWS_SESSION_TOKEN = None
  
s3_client = boto3.client(
  "s3",
  aws_access_key_id=AWS_ACCESS_KEY_ID,
  aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
  aws_session_token=AWS_SESSION_TOKEN,
)
books_df = pd.DataFrame(
  data={"Title": ["Book I", "Book II", "Book III"], "Price": [56.6, 59.87, 74.54]},
  columns=["Title", "Price"],
)


with io.StringIO() as csv_buffer:
  books_df.to_csv(csv_buffer, index=False)

  response = s3_client.put_object(
      Bucket=AWS_S3_BUCKET, Key="files/books.csv", Body=csv_buffer.getvalue()
  )

  status = response.get("ResponseMetadata", {}).get("HTTPStatusCode")

  if status == 200:
      print(f"Successful S3 put_object response. Status - {status}")
  else:
      print(f"Unsuccessful S3 put_object response. Status - {status}")
`,
        },
      ];
    case ConnectionType.Custom:
      return [
        {
          lang: "python",
          code: `# Importing os module 
import os

# Get connection fields from environment variables
${connection.fields
  .map(
    (f) => `${slugify(connection.slug, f.code)} = os.getenv("${slugify(
      connection.slug,
      f.code
    )}")
`
  )
  .join("")}
`,
        },
      ];
    default:
      return [];
  }
}

function updateFieldByCode(
  form: FormInstance<ConnectionForm>,
  field: FieldForm
) {
  const idx = form.formData.fields?.findIndex((f) => f.code === field.code);
  if (idx !== undefined && idx > -1) {
    updateFormField(form, idx, field);
  } else {
    form.setFieldValue("fields", [...(form.formData.fields ?? []), field]);
  }
}

const useConnectionFields = (
  form: FormInstance<ConnectionForm>,
  defaultFields: {
    code: string;
    name: string;
    secret?: boolean;
    value?: any;
  }[]
) => {
  const data = useMemo(() => {
    const result = defaultFields.map(
      (defaultField) =>
        form.formData.fields?.find((f) => f.code === defaultField.code) ??
        defaultField
    );
    return result.reduce((acc, current) => {
      acc[current.code] = current;
      return acc;
    }, {} as any);
    // We do not want this hook to be called if the reference of fieldTemplates changes
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [form.formData.fields]);

  const updateField = useCallback(
    (code: string, value: any) => {
      const field = data[code];
      updateFieldByCode(form, { ...field, value });
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [data, form.formData.fields]
  );

  useEffect(() => {
    // Update the form with the fields
    form.setFieldValue("fields", defaultFields);

    // Only set fields on first mount
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return [data, { updateField }] as const;
};

function PostgreSQLForm(props: { form: FormInstance<ConnectionForm> }) {
  const { form } = props;
  const { t } = useTranslation();

  const [{ db_name, host, port, user, password }, { updateField }] =
    useConnectionFields(form, [
      { code: "db_name", name: "DB Name" },
      { code: "host", name: "Host" },
      { code: "port", name: "Port" },
      { code: "user", name: "User" },
      { code: "password", secret: true, name: "Password" },
    ]);

  return (
    <>
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={db_name.value}
        name="db_name"
        label={t("Database name")}
        required
      />
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={host.value}
        name="host"
        label={t("Host")}
        required
      />
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={port.value}
        name="port"
        type="number"
        label={t("Port")}
        required
      />

      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={user.value}
        name="user"
        label={t("User")}
        required
      />

      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={password.value}
        name="password"
        type="password"
        label={t("Password")}
        required
      />
    </>
  );
}
function DHIS2Form(props: { form: FormInstance<ConnectionForm> }) {
  const { form } = props;
  const { t } = useTranslation();

  const [{ url, api_url, username, password }, { updateField }] =
    useConnectionFields(form, [
      { code: "url", name: "Instance" },
      { code: "api_url", name: "API Url" },
      { code: "username", name: "Username" },
      { code: "password", secret: true, name: "Password" },
    ]);

  return (
    <>
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={url.value}
        name="instance"
        type="url"
        label={t("Instance")}
        required
      />
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={api_url.value}
        name="api_url"
        type="url"
        label={t("API Url")}
      />

      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={username.value}
        name="username"
        label={t("Username")}
        required
      />

      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={password.value}
        name="password"
        type="password"
        label={t("Password")}
        required
      />
    </>
  );
}
function S3BucketForm(props: { form: FormInstance<ConnectionForm> }) {
  const { form } = props;
  const { t } = useTranslation();

  const [{ bucket_name, access_key_id, access_key_secret }, { updateField }] =
    useConnectionFields(form, [
      { code: "bucket_name", name: "Bucket name" },
      { code: "access_key_id", name: "Access key ID" },
      { code: "access_key_secret", name: "Secret access key", secret: true },
    ]);

  return (
    <>
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={bucket_name.value}
        name="bucket_name"
        label={t("Bucket name")}
        required
      />
      <div className="col-span-2">
        <Title level={6}>{t("Credentials")}</Title>
        <p className="text-sm text-gray-500">
          {t(
            "The credentials are required if the bucket cannot be accessed publicly."
          )}
        </p>
      </div>
      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={access_key_id.value}
        name="access_key_id"
        label={t("Access key ID")}
      />

      <Field
        onChange={(event) => updateField(event.target.name, event.target.value)}
        value={access_key_secret.value}
        name="access_key_secret"
        label={t("Secret access key")}
      />
    </>
  );
}
function GCSBucketForm(props: { form: FormInstance<ConnectionForm> }) {
  const { form } = props;
  const { t } = useTranslation();

  return null;
}

function CustomForm(props: { form: FormInstance<ConnectionForm> }) {
  const { form } = props;
  const { t } = useTranslation();

  return (
    <div className="col-span-2 space-y-3">
      <Title level={5}>{t("Fields")}</Title>
      <div className="max-h-80 space-y-2 overflow-y-auto py-px px-px">
        {form.formData.fields!.map((field, index) => (
          <div
            key={index}
            className="flex w-full items-center justify-end gap-2"
          >
            <Field
              className="flex-1"
              onChange={(event) =>
                updateFormField(form, index, {
                  code: event.target.value,
                  value: field.value,
                })
              }
              value={field.code}
              name={`code-${index}`}
              pattern="^[a-zA-Z0-9-_]*$"
              label={t("Name")}
              required
            />
            <Field
              name={`value-${index}`}
              label={t("Value")}
              onChange={(event) =>
                updateFormField(form, index, {
                  value: event.target.value,
                  code: field.code,
                })
              }
              value={field.value ?? ""}
              placeholder={t("Text value")}
              required
              className="flex-1"
            />
            <div className="mt-3 flex gap-2">
              <Checkbox
                id={`secret-${index}`}
                className="mt-1"
                name={`secret_${index}`}
                onChange={(event) =>
                  updateFormField(form, index, {
                    code: field.code,
                    secret: event.target.checked,
                  })
                }
                checked={field.secret ?? false}
                label={t("Secret")}
              />
              <button
                type="button"
                onClick={() =>
                  form.setFieldValue(
                    "fields",
                    form.formData.fields!.filter((_, i) => i !== index)
                  )
                }
              >
                <XMarkIcon className="h-4 w-4" />
              </button>
            </div>
          </div>
        ))}
      </div>

      <Button
        variant="white"
        size="sm"
        type="button"
        data-testid="add-field"
        onClick={() =>
          form.setFieldValue("fields", [...form.formData.fields!, {}])
        }
        leadingIcon={<PlusCircleIcon className="h-4 w-4" />}
      >
        {t("Add a field")}
      </Button>
    </div>
  );
}

export const CONNECTION_TYPES = {
  [ConnectionType.Postgresql]: {
    label: "PostgreSQL",
    color: "bg-blue-300",
    iconSrc: "/static/connector_postgresql/img/symbol.svg",
    Form: PostgreSQLForm,
  },
  [ConnectionType.Dhis2]: {
    label: "DHIS2 Instance",
    color: "bg-pink-300",
    iconSrc: "/static/connector_dhis2/img/symbol.svg",
    Form: DHIS2Form,
  },
  [ConnectionType.S3]: {
    label: "Amazon S3 Bucket",
    color: "bg-orange-200",
    iconSrc: "/static/connector_s3/img/symbol.svg",
    Form: S3BucketForm,
  },
  [ConnectionType.Gcs]: {
    label: "Google Cloud Bucket",
    color: "bg-blue-200",
    iconSrc: "/static/connector_gcs/img/symbol.svg",
    Form: GCSBucketForm,
  },
  [ConnectionType.Custom]: {
    label: "Custom",
    color: "bg-gray-200",
    iconSrc: null,
    Form: CustomForm,
  },
};

export function convertFieldsToInput(fields: FieldForm[]) {
  return fields.map((f) => ({
    code: f.code,
    secret: Boolean(f.secret),
    value: f.value ?? "",
  }));
}

export async function deleteConnection(connectionId: string) {
  const client = getApolloClient();
  const { data } = await client.mutate<
    DeleteConnectionMutation,
    DeleteConnectionMutationVariables
  >({
    mutation: gql`
      mutation DeleteConnection($input: DeleteConnectionInput!) {
        deleteConnection(input: $input) {
          success
          errors
        }
      }
    `,
    variables: { input: { id: connectionId } },
    update(cache) {
      const normalizedId = cache.identify({
        id: connectionId,
        __typename: "Connection",
      });

      cache.evict({ id: normalizedId });
      cache.gc();
    },
  });

  if (!data?.deleteConnection.success) {
    throw new Error("Impossible to delete connection");
  }
  return true;
}
