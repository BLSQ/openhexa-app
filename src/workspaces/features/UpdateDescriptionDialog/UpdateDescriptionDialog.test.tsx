import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  UpdateWorkspaceDocument,
  useUpdateWorkspaceMutation,
} from "workspaces/graphql/mutations.generated";
import { TestApp } from "core/helpers/testutils";
import WorkspaceDescriptionDialog from "./UpdateDescriptionDialog";
import { faker } from "@faker-js/faker";

jest.mock("workspaces/graphql/mutations.generated", () => ({
  ...jest.requireActual("workspaces/graphql/mutations.generated"),
  __esModule: true,
  useUpdateWorkspaceMutation: jest.fn().mockReturnValue([]),
}));

const WORKSPACE = {
  slug: faker.datatype.uuid(),
  name: faker.commerce.productName(),
};

const useUpdateWorkspaceMutationMock = useUpdateWorkspaceMutation as jest.Mock;

describe("EditWorkspaceDescriptionDialog", () => {
  const onClose = jest.fn();
  beforeEach(() => {
    useUpdateWorkspaceMutationMock.mockClear();
  });

  it("is not displayed ", async () => {
    const { container } = render(
      <WorkspaceDescriptionDialog
        workspace={WORKSPACE}
        open={false}
        onClose={() => {}}
      />
    );
    const dialog = await screen.queryByRole("dialog");
    expect(dialog).not.toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it("is displayed when open is true", async () => {
    const { container } = render(
      <TestApp mocks={[]}>
        <WorkspaceDescriptionDialog
          workspace={WORKSPACE}
          open={true}
          onClose={() => {}}
        />
      </TestApp>
    );

    const dialog = await screen.queryByRole("dialog");
    expect(dialog).toBeInTheDocument();
    expect(onClose).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();
  });

  it("Edit workspace description", async () => {
    const { useUpdateWorkspaceMutation } = jest.requireActual(
      "workspaces/graphql/mutations.generated"
    );
    useUpdateWorkspaceMutationMock.mockImplementation(
      useUpdateWorkspaceMutation
    );

    const user = userEvent.setup();
    const mocks = [
      {
        request: {
          query: UpdateWorkspaceDocument,
          variables: {
            input: {
              slug: WORKSPACE.slug,
            },
          },
        },
        result: {
          data: {
            update: {
              success: true,
              errors: [],
            },
          },
        },
      },
    ];

    const { container } = render(
      <TestApp mocks={mocks}>
        <WorkspaceDescriptionDialog
          workspace={WORKSPACE}
          open={true}
          onClose={() => {}}
        />
      </TestApp>
    );

    const descriptionInput = await screen.getByTestId("description");
    await user.clear(descriptionInput);

    const saveButton = screen.getByRole("button", { name: "Save" });
    await user.click(saveButton);
    await user.type(descriptionInput, "Description");
    expect(useUpdateWorkspaceMutationMock).toHaveBeenCalled();
  });
});
