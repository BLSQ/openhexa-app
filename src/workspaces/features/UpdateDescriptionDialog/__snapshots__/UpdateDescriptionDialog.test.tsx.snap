// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditWorkspaceDescriptionDialog is displayed when open is true 1`] = `
<div
  aria-hidden="true"
>
  <div
    style="position: fixed; top: 1px; left: 1px; width: 1px; height: 0px; padding: 0px; margin: -1px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px; display: none;"
  />
</div>
`;

exports[`EditWorkspaceDescriptionDialog is not displayed  1`] = `<div />`;
